pydantic/validators.py

```diff
@@ -656,9 +656,12 @@ def find_validators(  # noqa: C901 (ignore complexity)
     type_type = type_.__class__
     if type_type == ForwardRef or type_type == TypeVar:
         return
-    if type_ in NONE_TYPES:
-        yield none_validator
-        return
+    try:
+        if type_ in NONE_TYPES:
```

Comment:
I wonder if we should not write a custom set
```py
class IdentitySet(set):
    def __contains__(self, x):
        return any(x is v for v in self)
```
and have
```py
NONE_TYPES = IdentitySet((None, NoneType, Literal[None]))
```
to support all the
```py
whatever in NONE_TYPES
```
instead. @davidolrik WDYT?

Reply:
Yes, checking this correctly everywhere is a good idea. I tried a few variations and the custom class slowed down the benchmarks even in cython builds for me (I think this part is on the benchmark path).

I ended up just having a simple function that checks by identity, which does not seem to impact the cython benchmarks (At least I switched back and forth between the master and the new version, `make build`, `make benchmark-pydantic` and couldn't see any noticable impact.

In pure python, there might be a slight advantage to still using the try/except since the except is hardly ever called (only for non-hashable types which isn't the case very often, might even only be callables).
But testing using cython that advantage does not exist (cython probably can't optimize the try/except).
Here are a few timings of JUST the function (not pydantic benchmarks).

Using tuple and identity comparison (this is how it's implemented now)

```
In [10]: %%cython
    ...: from typing import Any, Literal
    ...: NoneType = None.__class__
    ...: 
    ...: 
    ...: NONE_TYPES = (None, NoneType, Literal[None])
    ...: 
    ...: 
    ...: def is_none_type(type_: Any) -> bool:
    ...:     for none_type in NONE_TYPES:
    ...:         if type_ is none_type:
    ...:             return True
    ...:     return False
    ...: 

In [11]: %timeit is_none_type(None)
30.2 ns ± 0.574 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)

In [12]: %timeit is_none_type(6)
33.9 ns ± 0.262 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)

In [13]: %timeit is_none_type({})
50.1 ns ± 11.8 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)
```

Using a set with try except:

```
In [6]: %%cython
   ...: from typing import Any, Literal
   ...: NoneType = None.__class__
   ...: 
   ...: 
   ...: NONE_TYPES: Set[Any] = {None, NoneType, Literal[None]}
   ...: 
   ...: 
   ...: def is_none_type(type_: Any) -> bool:
   ...:     try:
   ...:         return type_ in NONE_TYPES
   ...:     except TypeError:
   ...:         return False
   ...: 
   ...: 

In [7]: %timeit is_none_type(None)
50 ns ± 12.9 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)

In [8]: %timeit is_none_type(6)
40.9 ns ± 0.546 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)

In [9]: %timeit is_none_type({})
383 ns ± 121 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
```

For comparison: direct set "contains" (as it was before, but can't handle types like callable and so on), here we probably just save the function call:

```
In [14]: NONE_TYPES: Set[Any] = {None, NoneType, Literal[None]}

In [15]: %timeit None in NONE_TYPES
23.7 ns ± 0.201 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)

In [16]: %timeit 6 in NONE_TYPES
25 ns ± 0.0713 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)
```